jax/interpreters/xla.py

```diff
@@ -609,6 +610,13 @@ def _xla_callable(fun: lu.WrappedFun, device, backend, name, donated_invars, *ar
   log_priority = logging.WARNING if FLAGS.jax_log_compiles else logging.DEBUG
   logging.log(log_priority, "Compiling %s for args %s.", fun.__name__, abstract_args)
 
+  if nreps > 1:
+    print(f"The jitted function {fun.__name__} includes a pmap. Using "
+          "jit-of-pmap can lead to inefficient data movement, as the outer jit "
+          "does not preserve sharded data representations and instead collects "
+          "input and output arrays onto a single device. "
+          "Consider removing the outer jit unless you know what you're doing.")
```

Comment:
It looks like `warning.warn` will warn once per unique warning string:
```python
i = 0
def foo():
  global i
  warnings.warn("test")
  warnings.warn(f"test {i}")
  i += 1

# test.py:16: UserWarning: test
#   warnings.warn("test")
# test.py:17: UserWarning: test 0
#   warnings.warn(f"test {i}")
# test.py:17: UserWarning: test 1
#   warnings.warn(f"test {i}")
```
So you could use warnings here. There are of course cases where different functions have the same name, but I don't think printing the warning multiple times adds much value in this case. I'm slightly in favor of warnings because it seems like the right kind of signal and so you only get one warning even if you call the function with different shapes etc, but either is fine.

Reply:
Nice! TIL. Will switch to warnings.warn.

Reply:
This also let us add a test!